proposals/NNNN-isolated-default-values.md

```diff
@@ -0,0 +1,173 @@
+
+@MainActor func useDefault(
+  value: (Int, Int) = (requiresMainActor(), requiresAnotherActor()) // error!
+) {}
+```
+
+The above example is invalid because the default argument for `value` requires both `@MainActor` and `@AnotherActor`, but the caller can never satisfy both isolation requirements simultaneously.
+
+#### Closures
+
+Evaluating a closure literal itself can happen in any isolation domain; the actor isolation of a closure only applies when calling the closure. An actor-isolated closure enables the closure body to make calls within that isolation domain synchronously. For a closure literal in a default value expression that is not explicitly annotated with actor isolation, the inferred isolation of the closure is the union of isolation contexts of all callees in the closure body for synchronous calls. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+
+@MainActor func useDefaultClosure(
+  closure: () -> Void = {
+    requiresMainActor()
+  }
+) {}
+```
+
+The above `useDefaultClosure` function has a default argument value that is a closure literal. The closure body calls a `@MainActor`-isolated function synchronously, therefore the closure itself must be `@MainActor` isolated.
+
+Note that the inferred actor isolation of a closure literal can never be an actor instance based on the following two properties:
+
+1. To be isolated to an actor instance, a closure must either have its own (explicit) isolated parameter or capture an isolated parameter from its enclosing context.
+2. Closure literals in default arguments cannot capture values.
+
+### Enforcing default value isolation requirements
+
+#### Default argument values
+
+Isolation requirements for default argument expressions are enforced at the caller. If the caller is not in the required isolation domain, the default argument cannot be used and the argument must be specified explicitly. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+
+@MainActor func useDefault(value: Int = requiresMainActor()) { ... }
+
+@MainActor func mainActorCaller() {
+  useDefault() // okay
+}
+
+func nonisolatedCaller() async {
+  useDefault() // error
+
+  useDefault(value: await requiresMainActor()) // okay
+}
+```
+
+#### Stored property initial values
+
+Isolation requirements for default initializer expressions for stored properties apply in the body of initializers. If an `init` does not match the isolation of the initializer expression, the initialization of that stored property is not emitted at the beginning of the `init`. Instead, the stored property must be explicitly initialized in the body of the `init`. For example:
+
+```swift
+@MainActor func requiresMainActor() -> Int { ... }
+@AnotherActor func requiresAnotherActor() -> Int { ... }
+
+class C {
+  @MainActor var x1: Int = requiresMainActor()
+  @AnotherActor var x2: Int = requiresAnotherActor()
+
+  nonisolated init() {} // error
```

Comment:
Please explain why the error is here: neither `x1` nor `x2` are initialized because they have different isolations from the initializer.

Reply:
I explain the error in the prose just below the code; is that sufficient?

>In the above example, the no-parameter `nonisolated init()` is invalid, because it does not initialize `self.x1` and `self.x2`. Because the default initializer expressions require different actor isolation, those values are not used in the `nonisolated` initializer.

Reply:
... sure, that's sufficient. I suppose it's my personal style to prefer that each "error" in a code sample says why